### PR TITLE
Ground Propagator

### LINKED ISSUE #181: Add a C++ Ground Propagator Class
The orbit estimator needs to be able to propagate orbits sent from ground to the current time without too much computation a control cycle.

### PR DESCRIPTION
**Pull request by nhz2**
# Ground Propagator

Fixes #181 

### Summary of changes
- Added a GroundPropagator class to propagate orbits sent from ground.
- Added to docs of `Orbit::numgravcallsleft()` that it returns 0 if the Orbit is invalid.

### Ptest Effects
None

### Testing
Unit tests are in `test_groundpropagator.cpp`
Let me know if there is some edge case I'm missing that might break the estimator.

Under normal conditions, this estimator just propagates the most recently uplinked Orbit.

The main goal of the testing is to ensure there is no input that will break the estimator so it can't recover if ground does eventually uplink a good Orbit, and the sat is given enough time to propagate it to the current time.

Also the testing should ensure that too many Orbits getting uplinked won't overload the estimator and prevent it from making progress.

### Documentation Evidence
Inline documentation is sufficient.

### CODE REVIEW

File: include/orb/GroundPropagator.h [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+     * If catching_up is valid then catching_up is more recently input than current.
+     */
+    Orbit catching_up;
+
+    /** Orbit to catch up to current time once catching_up is done catching up.
+     * If to_catch_up is valid then catching_up is valid.
+     * If to_catch_up is valid then catching_up.numgravcallsleft()<to_catch_up.numgravcallsleft().
+     * If to_catch_up is valid then to_catch_up is more recently input than catching_up.
+     */
+    Orbit to_catch_up;
```
Comment (tanishqaggarwal) [inline include/orb/GroundPropagator.h:79]: If these items are frequently traded amongst themselves, can they be references or pointers? Orbit is an extremely heavy object.

File: include/orb/GroundPropagator.h [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+                //If that assumion is wrong, then the estimator isn't optimal, but it still works.
+            }
+        }
+        //start propagators
+        current.startpropagating(gps_time_ns,earth_rate_ecef);
+        catching_up.startpropagating(gps_time_ns,earth_rate_ecef);
+        to_catch_up.startpropagating(gps_time_ns,earth_rate_ecef);
+        //sort stuff
+        if (to_catch_up.valid() && to_catch_up.numgravcallsleft()<=catching_up.numgravcallsleft()){
+            catching_up= to_catch_up;
```
Comment (tanishqaggarwal) [inline include/orb/GroundPropagator.h:102]: Is there a copy assignment operator defined on Orbit? Might be necessary if Orbit contains pointers

File: include/orb/GroundPropagator.h [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+            to_catch_up= Orbit();
+        }
+    }
+
+    /** Return the total number of grav calls need to finish propagating all the orbits.
+     *
+     * grav calls: 0
+     */
+    int totalnumgravcallsleft(){
```
Comment (tanishqaggarwal) [inline include/orb/GroundPropagator.h]: Function can be marked const

File: include/orb/GroundPropagator.h [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+            to_catch_up= Orbit();
+        }
+    }
+
+    /** Do one grav call to move the propagator towards the current time if needed.
+     * The Orbit closest to finish propagating is propagated first.
+     * 
+     * grav calls: 1 or 0 if totalnumgravcallsleft() == 0 
+     */
+    void onegravcall(){
```
Comment (tanishqaggarwal) [inline include/orb/GroundPropagator.h]: To conform with functions everywhere else in FSW and GNC, could these functions be snake-case?

File: test/test_groundpropagator/test_groundpropagator.cpp [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+
+void test_basic_constructors() {
+    orb::GroundPropagator est;
+    orb::Orbit x= est.bestestimate();
+    TEST_ASSERT_FALSE(x.valid());
+    TEST_ASSERT_EQUAL_INT(est.totalnumgravcallsleft(),0);
+}
+
+
+/** Returns real fake orbits sent from ground.*/
```
Comment (tanishqaggarwal) [inline test/test_groundpropagator/test_groundpropagator.cpp:756]: What is a "real fake" orbit?

File: test/test_groundpropagator/test_groundpropagator.cpp [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+        break;
+    case 112:
+        //return an old orbit
+        r.startpropagating(r.nsgpstime()+1'000'000'000ULL,earth_rate_ecef);
+        r.finishpropagating();
+        return r;
+        break;
+    case 113:
+        //return an old orbit
+        r.startpropagating(r.nsgpstime()+20'000'000'000ULL,earth_rate_ecef);
```
Comment (tanishqaggarwal) [inline test/test_groundpropagator/test_groundpropagator.cpp:780]: How is overflow prevented in nsgpstime()?

File: test/test_groundpropagator/test_groundpropagator.cpp [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+        return r;
+        break;
+    default:
+        return orb::Orbit();
+        break;
+    }
+}
+
+
+void test_normal_use() {
```
Comment (tanishqaggarwal) [inline test/test_groundpropagator/test_groundpropagator.cpp:798]: The tests after this line are kind of hard to read. Consider adding more spacing and some header comments on each test describing: the purpose of the tests, the test steps, and how the steps demonstrate the purpose.

Comment (tanishqaggarwal) [reply test/test_groundpropagator/test_groundpropagator.cpp:798]: Consider breaking out common initializations into a TestFixture class, rather than copy-pasting the initializations into every test.

File: test/test_groundpropagator/test_groundpropagator.cpp [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+    lastsentorbit.finishpropagating();
+    best= est.bestestimate();
+    TEST_ASSERT_TRUE(best.valid());
+    TEST_ASSERT_FALSE(best.numgravcallsleft());
+    TEST_ASSERT_TRUE(best.nsgpstime()==lastsentorbit.nsgpstime());
+    PAN_TEST_ASSERT_LIN_3VECT_WITHIN(1E-6, best.vecef(), lastsentorbit.vecef());
+    PAN_TEST_ASSERT_LIN_3VECT_WITHIN(1E-3, best.recef(), lastsentorbit.recef());
+}
+
+void test_edge_cases() {
```
Comment (tanishqaggarwal) [inline test/test_groundpropagator/test_groundpropagator.cpp]: Could you describe what each of these edge cases are, in comments?

**Review by tanishqaggarwal — CHANGES_REQUESTED** [review]
My main concern is with test readability. The propagator interface looks pretty good and readily integrable into FSW.

File: include/orb/GroundPropagator.h (hunk 1 above)
Comment (kylekrol) [reply include/orb/GroundPropagator.h:79]: While this may be a valid concern I feel like the additional complexity in creating some buffer for orbit objects and then passing around pointers to it may not be worth the extra effort. The relative overhead of a call to `memcpy` compared to the actual propagation is so minimal as well.

File: include/orb/GroundPropagator.h (hunk 2 above)
Comment (kylekrol) [reply include/orb/GroundPropagator.h:102]: I don't there are any internal pointers. However, it may be possible to write a more efficient copy assignment operator/constructor if we don't have to copy all of the internal fields into the new object. For example, if we don't need copy assignment while an orbit is propagating and only need to copy over `r`, `v`, and `t` that could save some overhead.

File: include/orb/GroundPropagator.h [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+        return current.numgravcallsleft()+catching_up.numgravcallsleft()+to_catch_up.numgravcallsleft();
+    }
+
+    /** Returns a reference to the best estimate of the Orbit.
+     * 
+     * grav calls: 0
+     */
+    Orbit bestestimate() const{
+        return current;
+    }
```
Comment (kylekrol) [inline include/orb/GroundPropagator.h]: How exactly is this going to integrate with the orbit filter, or are we just going to propagate and not filter orbits received from the ground (which would totally be fine)?

**Review by kylekrol — COMMENTED** [review]
Overall, the propegation logic and everything looks great. I am wondering, however, if we'd want the `bestestimate` function to ever return an orbit that wasn't caught up to the present. I'm not sure if this will ever be useful information for the rest of gnc/fsw/ Thoughts?

File: test/test_groundpropagator/test_groundpropagator.cpp (hunk 6 above)
Comment (nhz2) [reply test/test_groundpropagator/test_groundpropagator.cpp:780]: There is no overflow protection right now, but nsgpstime is a 64 bit unsigned int, so it won't overflow for hundreds of years.

File: include/orb/GroundPropagator.h (hunk 9 above)
Comment (nhz2) [reply include/orb/GroundPropagator.h]: Yeah, the main orbit estimator is going to have a ground propagator for both the self and target orbits. These will used to initialize/reset the Kalman filters, and as a backup incase the GPS or CDGPS isn't giving good data and/or the Kalman filter diverges. There will be a state field flag to just use the ground propagators and ignore GPS, which will also be helpful for testing.

File: include/orb/GroundPropagator.h (hunk 2 above)
Comment (nhz2) [reply include/orb/GroundPropagator.h:102]: Yeah, there are no internal pointers.

File: include/orb/GroundPropagator.h (hunk 1 above)
Comment (nhz2) [reply include/orb/GroundPropagator.h:79]: #183 I made an issue to try this in a future PR if needed.

**Comment by nhz2** [comment]
> Overall, the propegation logic and everything looks great. I am wondering, however, if we'd want the `bestestimate` function to ever return an orbit that wasn't caught up to the present. I'm not sure if this will ever be useful information for the rest of gnc/fsw/ Thoughts?

It probably won't be useful for the rest of GNC if the `best_estimate` isn't caught up. Hopefully this will only happen on startup, or if the GPS time and teensy clock get off by a huge amount (like more than 1 second) for some reason. GNC and flight software can check by calling `est.best_estimate().numgravcallsleft()` and `est.best_estimate().valid()` I'm not sure, but hopefully that will just be inlined and won't actually copy the current Orbit into a temporary object and then call numgravcallsleft().

File: include/orb/GroundPropagator.h [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+        return current.numgravcallsleft()+catching_up.numgravcallsleft()+to_catch_up.numgravcallsleft();
+    }
+
+    /** Returns a reference to the best estimate of the Orbit.
+     * 
+     * grav calls: 0
+     */
+    Orbit best_estimate() const{
+        return current;
+    }
```
Comment (kylekrol) [inline include/orb/GroundPropagator.h:158]: Return an `Orbit const &` here may more explicitly prevent copying (per your comment above). The use also have the the option to generate a copy themselves from the reference like `Orbit orbit = propegator.best_estimate();`

File: test/test_groundpropagator/test_groundpropagator.cpp (hunk 6 above)
Comment (tanishqaggarwal) [reply test/test_groundpropagator/test_groundpropagator.cpp:780]: Ah, I see, nanoseconds in 64 bits is ok.

File: include/orb/GroundPropagator.h (hunk 1 above)
Comment (nhz2) [reply include/orb/GroundPropagator.h:79]: I'm closing the issue because coping will only happen rarely when an Orbit gets caught up or a new Orbit gets uplinked from ground, not every control cycle.

**Review by kylekrol — APPROVED** [review]

**Review by tanishqaggarwal — APPROVED** [review]
I don't have the time to read through all of the tests this week, but I know they're probably correct and in the interests of moving progress forward, approving this PR
